File: src/main/java/com/adil/nsealerts/TradeRules.java
package com.adil.nsealerts;

/**
 * Pure exit-rule decision logic. No I/O, no Spring — a function of trade state,
 * current price, and elapsed time only.
 *
 * Used by both UpstoxTradeService (real orders) and ShadowTradeService /
 * the backtest simulator (paper positions), so the rules that decide when
 * money moves and the rules used to evaluate the strategy can never drift
 * apart from each other.
 *
 * Check order matters and mirrors the original inline logic in
 * UpstoxTradeService.monitorPosition(): time exit first (always wins),
 * then hard stop, then partial exit, then trailing stop.
 */
public final class TradeRules {

    private TradeRules() {}

    public enum Action { HOLD, PARTIAL_EXIT, TRAIL_STOP, HARD_STOP, TIME_EXIT }

    public record Params(
            double targetPct,
            double stopLossPct,
            double trailPct,
            int timeExitMinutes) {}

    public record Decision(Action action, String reason) {
        static Decision hold() { return new Decision(Action.HOLD, ""); }
    }

    /**
     * @param gainPct         (ltp - entryPrice) / entryPrice * 100, at this tick
     * @param maxGainPct      highest gainPct observed since entry, INCLUDING this tick
     * @param partialExitDone whether the target-based partial exit has already fired
     * @param ageMinutes      whole minutes elapsed since entry
     */
    public static Decision decide(Params p, double gainPct, double maxGainPct,
                                   boolean partialExitDone, long ageMinutes) {
        if (ageMinutes >= p.timeExitMinutes()) {
            return new Decision(Action.TIME_EXIT,
                    String.format("Time exit (%d min)", p.timeExitMinutes()));
        }
        if (gainPct <= -p.stopLossPct()) {
            return new Decision(Action.HARD_STOP,
                    String.format("Stop-loss (%.2f%%)", gainPct));
        }
        if (!partialExitDone && gainPct >= p.targetPct()) {
            return new Decision(Action.PARTIAL_EXIT,
                    String.format("Partial exit at target (+%.2f%%)", gainPct));
        }
        if (partialExitDone && maxGainPct >= p.trailPct()) {
            double trailTrigger = maxGainPct - p.trailPct();
            if (gainPct <= trailTrigger) {
                return new Decision(Action.TRAIL_STOP,
                        String.format("Trail stop (peak:+%.2f%% -> now:+%.2f%%)", maxGainPct, gainPct));
            }
        }
        return Decision.hold();
    }
}

File: backtest/tradeRules.js
'use strict';

/**
 * Mirrors TradeRules.java's decide() exactly (same thresholds, same check order:
 * time exit, then hard stop, then partial exit, then trailing stop). Keep the two
 * in sync by hand when either changes — there is no automated cross-language check.
 * See src/main/java/com/adil/nsealerts/UpstoxTradeService.java for how the Java
 * side wires this into live trading and shadow-mode simulation.
 */

const Action = Object.freeze({
  HOLD: 'HOLD',
  PARTIAL_EXIT: 'PARTIAL_EXIT',
  TRAIL_STOP: 'TRAIL_STOP',
  HARD_STOP: 'HARD_STOP',
  TIME_EXIT: 'TIME_EXIT',
});

/**
 * @param {{targetPct:number, stopLossPct:number, trailPct:number, timeExitMinutes:number}} params
 * @param {number} gainPct         (price - entryPrice) / entryPrice * 100, at this tick
 * @param {number} maxGainPct      highest gainPct observed since entry, INCLUDING this tick
 * @param {boolean} partialExitDone whether the target-based partial exit has already fired
 * @param {number} ageMinutes      whole minutes elapsed since entry
 */
function decide(params, gainPct, maxGainPct, partialExitDone, ageMinutes) {
  const { targetPct, stopLossPct, trailPct, timeExitMinutes } = params;

  if (ageMinutes >= timeExitMinutes) {
    return { action: Action.TIME_EXIT, reason: `Time exit (${timeExitMinutes} min)` };
  }
  if (gainPct <= -stopLossPct) {
    return { action: Action.HARD_STOP, reason: `Stop-loss (${gainPct.toFixed(2)}%)` };
  }
  if (!partialExitDone && gainPct >= targetPct) {
    return { action: Action.PARTIAL_EXIT, reason: `Partial exit at target (+${gainPct.toFixed(2)}%)` };
  }
  if (partialExitDone && maxGainPct >= trailPct) {
    const trailTrigger = maxGainPct - trailPct;
    if (gainPct <= trailTrigger) {
      return {
        action: Action.TRAIL_STOP,
        reason: `Trail stop (peak:+${maxGainPct.toFixed(2)}% -> now:+${gainPct.toFixed(2)}%)`,
      };
    }
  }
  return { action: Action.HOLD, reason: '' };
}

module.exports = { Action, decide };
